sycl/include/sycl/ext/oneapi/experimental/chunk.hpp

```diff
@@ -0,0 +1,233 @@
+#else
+    return linear_id_type(0);
+#endif
+  }
+
+  linear_id_type get_local_linear_range() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return static_cast<linear_id_type>(get_local_range()[0]);
+#else
+    return linear_id_type(0);
+#endif
+  }
+
+  bool leader() const {
+#ifdef __SYCL_DEVICE_ONLY__
+    return get_local_linear_id() == 0;
+#else
+    return linear_id_type(0);
+#endif
+  }
+
+protected:
+  static constexpr size_t chunk_size = ChunkSize;
+
+#if defined(__SYCL_DEVICE_ONLY__) && defined(__NVPTX__)
+  sub_group_mask Mask;
+
+  chunk(ext::oneapi::sub_group_mask mask) : Mask(mask) {}
+
+  ext::oneapi::sub_group_mask getMask() const { return Mask; }
+#else
+  chunk() {}
+
+#ifdef __SYCL_DEVICE_ONLY__
+  ext::oneapi::sub_group_mask getMask() const {
+    ext::oneapi::sub_group_mask::BitsType MaskBits{0};
+    MaskBits = ~MaskBits;
+    MaskBits <<= ext::oneapi::sub_group_mask::max_bits - ChunkSize;
+    MaskBits >>=
+        ext::oneapi::sub_group_mask::max_bits -
+        (((__spirv_SubgroupLocalInvocationId() / ChunkSize) + 1) * ChunkSize);
+    return sycl::detail::Builder::createSubGroupMask<
+        ext::oneapi::sub_group_mask>(MaskBits, __spirv_SubgroupMaxSize());
+  }
+#endif
+#endif
+
+  friend chunk<ChunkSize, ParentGroup>
+  chunked_partition<ChunkSize, ParentGroup>(ParentGroup parent);
+
+  friend sub_group_mask sycl::detail::GetMask<chunk<ChunkSize, ParentGroup>>(
+      chunk<ChunkSize, ParentGroup> Group);
+
+  template <size_t OtherChunkSize, typename OtherParentGroup>
+  friend class chunk;
+};
+
+// Chunked partition implementation
+template <size_t ChunkSize, typename ParentGroup>
+inline std::enable_if_t<std::is_same_v<ParentGroup, sycl::sub_group> ||
+                            is_chunk_v<ParentGroup>,
+                        chunk<ChunkSize, ParentGroup>>
+chunked_partition([[maybe_unused]] ParentGroup parent) {
+  static_assert((ChunkSize & (ChunkSize - size_t{1})) == size_t{0},
+                "ChunkSize must be a power of 2.");
+#ifdef __SYCL_DEVICE_ONLY__
+  // sync all work-items in parent group before partitioning
+  sycl::group_barrier(parent);
+
+#if defined(__NVPTX__)
+  uint32_t loc_id = parent.get_local_linear_id();
+  uint32_t loc_size = parent.get_local_linear_range();
+  uint32_t bits = ChunkSize == 32 ? 0xffffffff
+                                  : ((1 << ChunkSize) - 1)
+                                        << ((loc_id / ChunkSize) * ChunkSize);
+
+  return chunk<ChunkSize, ParentGroup>(
+      sycl::detail::Builder::createSubGroupMask<ext::oneapi::sub_group_mask>(
+          bits, loc_size));
+#else
+  return chunk<ChunkSize, ParentGroup>();
+#endif
+#else
+  return chunk<ChunkSize, ParentGroup>();
+#endif
+}
+
+// Type traits
+template <size_t ChunkSize, typename ParentGroup>
+struct is_user_constructed_group<chunk<ChunkSize, ParentGroup>>
+    : std::true_type {};
+
+template <size_t ChunkSize, typename ParentGroup>
+struct is_chunk<chunk<ChunkSize, ParentGroup>> : std::true_type {};
```

Comment:
This isn't part of the specification, so I think it needs to be in a `detail::` namespace.

Reply:
Good catch! In fact, we had them defined twice and there were some leftover traits that could be removed.